### PR TITLE
api: Added more bucket ops to the api

### PR DESCRIPTION
**Pull request by thephred**
cli: Added a few more bucket ops to the cli

### CODE REVIEW

File: pkg/cephmgr/rgw/bucket.go [hunk 1]
```diff
@@ -49,53 +82,136 @@ func GetBucketStats(context *clusterd.Context, getClusterInfo func() (*mon.Clust
...
+		buckets = append(buckets, model.ObjectBucket{Name: bucket, ObjectBucketMetadata: model.ObjectBucketMetadata{Owner: metadata.Owner, CreatedAt: metadata.CreatedAt}, ObjectBucketStats: stat})
+	}
+
+	return buckets, nil
+}
+
+func GetBucket(context *clusterd.Context, bucket string, getClusterInfo func() (*mon.ClusterInfo, error)) (*model.ObjectBucket, int, error) {
+	stat, err := GetBucketStats(context, bucket, getClusterInfo)
+	if err != nil {
+		if err.Error() == "not found" {
```
Comment (travisn) [inline pkg/cephmgr/rgw/bucket.go]: is there a way we can avoid the error checking with a string compare?

File: pkg/cephmgr/rgw/bucket.go [hunk 2]
```diff
@@ -49,53 +82,136 @@ func GetBucketStats(context *clusterd.Context, getClusterInfo func() (*mon.Clust
...
+
+	if strings.Contains(result, "could not get bucket info for bucket=") {
+		return fmt.Errorf("bucket not found")
+	}
+
+	return fmt.Errorf("failed to delete bucket: %+v", err)
+}
+
+func GetBucketACL(context *clusterd.Context, bucketName string, getClusterInfo func() (*mon.ClusterInfo, error)) (*model.ObjectBucketACL, error) {
+	return nil, fmt.Errorf("Not yet implemented")
```
Comment (travisn) [inline pkg/cephmgr/rgw/bucket.go]: i think we either need to remove the ACL placeholders in this PR, or finish the implementation before the PR is merged

File: pkg/cephmgr/rgw/bucket.go [hunk 3]
```diff
@@ -49,53 +82,136 @@ func GetBucketStats(context *clusterd.Context, getClusterInfo func() (*mon.Clust
...
 		}
+		return nil, RGWErrorUnknown, err
+	}
 
-		buckets = append(buckets, model.ObjectBucket{Name: bucket, Owner: s.Data.Owner, CreatedAt: createdAt, ObjectBucketStats: stat})
+	return &model.ObjectBucket{Name: bucket, ObjectBucketMetadata: model.ObjectBucketMetadata{Owner: metadata.Owner, CreatedAt: metadata.CreatedAt}, ObjectBucketStats: *stat}, RGWErrorNone, nil
+}
+
+// FIXME: The "bucket create" radowsgw-admin command does not exist yet
+func CreateBucket(context *clusterd.Context, bucket model.ObjectBucket, getClusterInfo func() (*mon.ClusterInfo, error)) (*model.ObjectBucket, error) {
```
Comment (travisn) [inline pkg/cephmgr/rgw/bucket.go]: similar to ACLs, i don't think it makes sense to expose the create command without an implementation

File: pkg/rook/client/object.go [hunk 4]
```diff
@@ -148,7 +229,7 @@ func (c *RookNetworkRestClient) UpdateObjectUser(user model.ObjectUser) (*model.
 func (c *RookNetworkRestClient) DeleteObjectUser(id string) error {
 	query := path.Join(objectStoreQueryName, usersQueryName, id)
 	_, err := c.DoDelete(query)
-	if err != nil {
+	if err != nil && !IsHttpStatusCode(err, http.StatusNoContent) {
```
Comment (travisn) [inline pkg/rook/client/object.go]: StatusNoContent is the success response?

Comment (thephred) [inline pkg/rook/client/object.go]: Yes

File: cmd/rook/object/bucket.go [hunk 5]
```diff
@@ -81,3 +91,96 @@ func listBuckets(c client.RookRestClient) (string, error) {
 	w.Flush()
 	return buffer.String(), nil
 }
+
+var bucketGetCmd = &cobra.Command{
+	Use:   "get [Bucket Name]",
```
Comment (thephred) [inline cmd/rook/object/bucket.go]: good call

File: cmd/rook/object/bucket.go [hunk 6]
```diff
@@ -81,3 +91,96 @@ func listBuckets(c client.RookRestClient) (string, error) {
...
+
+	if len(args) == 0 {
+		return fmt.Errorf("Missing required argument Bucket Name")
+	}
+
+	if len(args) > 1 {
+		return fmt.Errorf("Too many arguments")
+	}
+
+	if err := flags.VerifyRequiredFlags(cmd, []string{}); err != nil {
```
Comment (thephred) [inline cmd/rook/object/bucket.go]: will do

File: pkg/api/object.go [hunk 7]
```diff
@@ -221,11 +220,128 @@ func (h *Handler) Listbuckets(w http.ResponseWriter, r *http.Request) {
...
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	FormatJsonResponse(w, createdBucket)
+}
+
+// GetBucket gets the bucket from the object store in this cluster.
+// GET
+// /objectstore/users/{USER_ID}
```
Comment (thephred) [inline pkg/api/object.go]: good catch

File: pkg/api/object.go [hunk 8]
```diff
@@ -221,11 +220,128 @@ func (h *Handler) Listbuckets(w http.ResponseWriter, r *http.Request) {
...
+func (h *Handler) GetBucket(w http.ResponseWriter, r *http.Request) {
+	bucketName := mux.Vars(r)["bucketName"]
+
+	user, rgwError, err := rgw.GetBucket(h.context, bucketName, h.config.ClusterHandler.GetClusterInfo)
+	if err != nil {
+		logger.Errorf("Error getting user (%s): %+v", bucketName, err)
```
Comment (thephred) [inline pkg/api/object.go]: right, bucket

File: pkg/api/object.go [hunk 9]
```diff
@@ -221,11 +220,128 @@ func (h *Handler) Listbuckets(w http.ResponseWriter, r *http.Request) {
...
+// GetBucketACL gets the ACL of the bucket in the object store in this cluster.
+// GET
+// /objectstore/bucketACLs/{BUCKET_NAME}
+func (h *Handler) GetBucketACL(w http.ResponseWriter, r *http.Request) {
+	bucketName := mux.Vars(r)["bucketName"]
+
+	acl, err := rgw.GetBucketACL(h.context, bucketName, h.config.ClusterHandler.GetClusterInfo)
+	if err != nil {
+		logger.Errorf("Error getting bucket ACL: %+v", err)
+		w.WriteHeader(http.StatusInternalServerError)
```
Comment (thephred) [inline pkg/api/object.go]: At this point there is no mechanism for distinguishing between not found and other errors. This should change once we have an the code written that backs this in ceph.

File: pkg/api/object.go [hunk 10]
```diff
@@ -221,11 +220,128 @@ func (h *Handler) Listbuckets(w http.ResponseWriter, r *http.Request) {
...
+		w.Write([]byte(err.Error()))
```
Comment (thephred) [inline pkg/api/object.go]: I could be yes.

File: pkg/api/object.go [hunk 11]
```diff
@@ -221,11 +220,128 @@ func (h *Handler) Listbuckets(w http.ResponseWriter, r *http.Request) {
...
+		logger.Errorf("Error parsing bucket ACL: %+v", err)
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(err.Error()))
+		return
+	}
+
+	err = rgw.SetBucketACL(h.context, bucketName, acl, h.config.ClusterHandler.GetClusterInfo)
+	if err != nil {
+		logger.Errorf("Error getting bucket ACL: %+v", err)
 		w.WriteHeader(http.StatusInternalServerError)
```
Comment (thephred) [inline pkg/api/object.go]: Not yet, see above.

File: pkg/api/routes.go [hunk 12]
```diff
@@ -123,7 +123,37 @@ func (h *Handler) GetRoutes() []Route {
 			"ListBuckets",
 			"GET",
 			"/objectstore/buckets",
-			h.Listbuckets,
+			h.ListBuckets,
+		},
+		{
+			"CreateBucket",
+			"POST",
+			"/objectstore/buckets",
```
Comment (thephred) [inline pkg/api/routes.go]: It looks like there is debate one which way is better and both are in use.  I don't think either is right or wrong but personally I think the plural form makes more sense.  http://www.restapitutorial.com/lessons/restfulresourcenaming.html

File: pkg/cephmgr/rgw/bucket.go [hunk 13]
```diff
@@ -33,7 +34,39 @@ type rgwBucketStats struct {
 	}
 }
 
-func GetBucketStats(context *clusterd.Context, getClusterInfo func() (*mon.ClusterInfo, error)) (map[string]model.ObjectBucketStats, error) {
+func bucketStatsFromRGW(rgwStat rgwBucketStats) model.ObjectBucketStats {
+	stat := model.ObjectBucketStats{Size: 0, NumberOfObjects: 0}
```
Comment (thephred) [inline pkg/cephmgr/rgw/bucket.go]: Seems like it makes intent clear.

File: pkg/api/routes.go [hunk 14]
```diff
@@ -123,7 +123,37 @@ func (h *Handler) GetRoutes() []Route {
...
+		{
+			"GetBucket",
+			"GET",
+			"/objectstore/buckets/{bucketName}",
+			h.GetBucket,
+		},
+		{
+			"GetBucketACL",
+			"GET",
+			"/objectstore/bucketACLs/{bucketName}",
```
Comment (thephred) [inline pkg/api/routes.go]: as discussed I will change this to /objectstore/buckets/{bucketName}/acl

File: pkg/cephmgr/rgw/bucket.go [hunk 15]
```diff
@@ -33,7 +34,39 @@ type rgwBucketStats struct {
 	}
 }
 
-func GetBucketStats(context *clusterd.Context, getClusterInfo func() (*mon.ClusterInfo, error)) (map[string]model.ObjectBucketStats, error) {
+func bucketStatsFromRGW(rgwStat rgwBucketStats) model.ObjectBucketStats {
```
Comment (thephred) [inline pkg/cephmgr/rgw/bucket.go]: changed

File: pkg/cephmgr/rgw/bucket.go (hunk 1 above)
Comment (thephred) [inline pkg/cephmgr/rgw/bucket.go]: will fix

File: cmd/rook/object/bucket.go (hunk 6 above)
Comment (travisn) [inline cmd/rook/object/bucket.go]: looks like one more instance of VerifyRequiredFlags to remove

File: pkg/api/object.go [hunk 16]
```diff
@@ -221,11 +220,47 @@ func (h *Handler) Listbuckets(w http.ResponseWriter, r *http.Request) {
...
-	if util.IsEtcdKeyNotFound(err) {
-		w.WriteHeader(http.StatusNotFound)
-	} else {
+// GetBucket gets the bucket from the object store in this cluster.
+// GET
+// /objectstore/buckets/{USER_ID}
+func (h *Handler) GetBucket(w http.ResponseWriter, r *http.Request) {
+	bucketName := mux.Vars(r)["bucketName"]
+
+	user, rgwError, err := rgw.GetBucket(h.context, bucketName, h.config.ClusterHandler.GetClusterInfo)
```
Comment (jbw976) [inline pkg/api/object.go]: it looks like `bucketname` is required.  If it's not part of the HTTP request, should we return `BadRequest`?

File: pkg/api/object.go [hunk 17]
```diff
@@ -221,11 +220,47 @@ func (h *Handler) Listbuckets(w http.ResponseWriter, r *http.Request) {
...
 }
 
-func handleConnectionLookupFailure(err error, action string, w http.ResponseWriter) {
-	logger.Errorf("failed to get %s: %+v", action, err)
-	if util.IsEtcdKeyNotFound(err) {
-		w.WriteHeader(http.StatusNotFound)
-	} else {
+// GetBucket gets the bucket from the object store in this cluster.
+// GET
+// /objectstore/buckets/{USER_ID}
```
Comment (jbw976) [inline pkg/api/object.go]: should the `bucketname` query param be added to this comment? is `{USER_ID}` needed or is it a typo?

File: pkg/api/object.go [hunk 18]
```diff
@@ -221,11 +220,47 @@ func (h *Handler) Listbuckets(w http.ResponseWriter, r *http.Request) {
...
+		}
+		return
+	}
+
+	FormatJsonResponse(w, user)
+}
+
+// DeleteBucket deletes the bucket in the object store in this cluster.
+// DELETE
+// /objectstore/buckets/{BUCKET_NAME}
```
Comment (jbw976) [inline pkg/api/object.go]: does this operate in the context of a particular user? or can this delete anyone's buckets?

File: pkg/api/object.go [hunk 19]
```diff
@@ -221,11 +220,47 @@ func (h *Handler) Listbuckets(w http.ResponseWriter, r *http.Request) {
...
+func (h *Handler) DeleteBucket(w http.ResponseWriter, r *http.Request) {
+	bucketName := mux.Vars(r)["bucketName"]
+
+	purgeParams, found := r.URL.Query()["purge"]
+	purge := found && len(purgeParams) == 1 && purgeParams[0] == "true"
+
+	err := rgw.DeleteBucket(h.context, bucketName, purge, h.config.ClusterHandler.GetClusterInfo)
+	if err != nil {
+		if err.Error() == "bucket not found" {
```
Comment (jbw976) [inline pkg/api/object.go]: why does this one not use `rgw.RGWErrorNotFound` like the `GetBucket` API does?

File: pkg/model/object.go [hunk 20]
```diff
@@ -55,3 +59,21 @@ func (slice ObjectBuckets) Less(i, j int) bool {
 func (slice ObjectBuckets) Swap(i, j int) {
 	slice[i], slice[j] = slice[j], slice[i]
 }
+
+type ObjectBucketGrant struct {
```
Comment (jbw976) [inline pkg/model/object.go]: do these types not require any serialization annotations?

File: pkg/api/object.go (hunk 17 above)
Comment (thephred) [inline pkg/api/object.go]: typo

File: pkg/api/object.go (hunk 16 above)
Comment (thephred) [inline pkg/api/object.go]: This gets caught at the routing level and will actually 404 if it isn't passed.

File: pkg/api/object.go (hunk 18 above)
Comment (thephred) [inline pkg/api/object.go]: anyone's buckets

File: pkg/api/object.go (hunk 19 above)
Comment (thephred) [inline pkg/api/object.go]: No reason, I'll change it.

File: pkg/model/object.go (hunk 20 above)
Comment (thephred) [inline pkg/model/object.go]: This should have been removed. I'll clean that up.

**Pull request merged**